### PR TITLE
fix: release fix 2.38: Update prescription linked model sync behaviour

### PR DESCRIPTION
**Pull request by tcodling**
### Changes

After bug + discussion with Michael we decided that we would treat ongoing prescriptions as 'sensitive' so this pr adjusts the logic accordingly.

### Deploys

- [x] **Deploy to Tamanu Internal** 

### Tests

- [ ] **Run E2E Tests** 

### Remember to...

- ...write or update tests
- ...add UI screenshots and **testing notes** to the Linear issue
- ...add any **manual upgrade steps** to the Linear issue
- ...update the [config reference](https://beyond-essential.slab.com/posts/reference-config-file-0c70ukly), [settings reference](https://beyond-essential.slab.com/posts/reference-settings-0blw1x2q), or any [relevant runbook(s)](https://beyond-essential.slab.com/topics/runbooks-bs04ml6c)
- ...call out additions or changes to **config files** for the deployment team to take note of

### CODE REVIEW

File: packages/central-server/__tests__/sync/CentralSyncManager.test.js [hunk 1]
```diff
@@ -3980,37 +3980,27 @@ describe('CentralSyncManager', () => {
         expect(updatedEncounterIds).not.toContain(encounter.id);
       });
 
-      it('will sync prescriptions linked through patient_ongoing_prescriptions if marked for sync', async () => {
-        const testPatient1 = await models.Patient.create(fake(models.Patient));
-        const testPatient2 = await models.Patient.create(fake(models.Patient));
+      it('will not sync prescriptions linked through patient_ongoing_prescriptions from a sensitive facility', async () => {
```
Comment (rohan-bes) [inline packages/central-server/__tests__/sync/CentralSyncManager.test.js:3983]: At the moment I don't think this test is quite doing what we want it to do. 

I think what we want is:
1. Create a `patient`
2. Create an `encounter` at `sensitiveFacility` for the `patient`
3. Create a `prescription`
4. Create an `encounterPrescription` for the `encounter`/`prescription`
5. Create a `patientOngoingPrescription` for the  `patient`/`prescription`
6. Run a sync for the `nonSensitiveFacility`
7. Check to ensure the sync does not contain the `encounter`/`prescription`/`encounterPrescription`/`patientOngoingPrescription`

**Review by rohan-bes — CHANGES_REQUESTED** [review]
Nice stuff! I think the test needs a bit of an update :+1:

File: packages/database/src/models/PatientOngoingPrescription.ts [hunk 2]
```diff
@@ -35,7 +35,23 @@ export class PatientOngoingPrescription extends Model {
...
-    return buildPatientLinkedLookupFilter(this);
+    return {
+      select: buildSyncLookupSelect(this, {
+        patientId: 'patient_ongoing_prescriptions.patient_id',
+        facilityId: `
+          CASE
+            WHEN facilities.is_sensitive = TRUE THEN facilities.id
+            ELSE NULL
+          END
+        `,
```
[the remark this answers is not in the document]
Comment (tcodling) [reply packages/database/src/models/PatientOngoingPrescription.ts:46]: For sure!

File: packages/central-server/__tests__/sync/CentralSyncManager.test.js (hunk 1 above)
Comment (tcodling) [reply packages/central-server/__tests__/sync/CentralSyncManager.test.js:3983]: Oh yea i didnt even change the expect 🤦

**Review by rohan-bes — APPROVED** [review]
Awesome thank u!

File: packages/database/src/models/PatientOngoingPrescription.ts [hunk 3]
```diff
@@ -35,7 +36,18 @@ export class PatientOngoingPrescription extends Model {
...
+        patientId: 'patient_ongoing_prescriptions.patient_id',
+        facilityId: ADD_SENSITIVE_FACILITY_ID_IF_APPLICABLE,
+      }),
+      joins: `
+        LEFT JOIN encounter_prescriptions ON patient_ongoing_prescriptions.prescription_id = encounter_prescriptions.prescription_id
+        LEFT JOIN encounters ON encounter_prescriptions.encounter_id = encounters.id
+        LEFT JOIN locations ON encounters.location_id = locations.id
+        LEFT JOIN facilities ON locations.facility_id = facilities.id
+      `,
+    };
```
[the remark this answers is not in the document]
Comment (tcodling) [reply packages/database/src/models/PatientOngoingPrescription.ts:50]: @rohan-bes This comment has scared me a bit 😆 Might write a few more tests just for my own confidence

[the remark this answers is not in the document]
Comment (rohan-bes) [reply packages/database/src/models/PatientOngoingPrescription.ts:50]: Yeah damn good catch AI, this is a genuine issue :\ and the join to get a single row is kinda awkward...

I would say the same issue exists for `prescription` as well... and it pre-dates your change

[the remark this answers is not in the document]
Comment (tcodling) [reply packages/database/src/models/PatientOngoingPrescription.ts:50]: Hmm so resolve i would just need to make sure it resolves to one row with a facility id?

File: packages/database/src/models/MedicationAdministrationRecord.ts [hunk 4]
```diff
@@ -558,10 +559,20 @@ export class MedicationAdministrationRecord extends Model {
 
   static buildSyncLookupQueryDetails() {
     return {
-      select: buildEncounterLinkedLookupSelect(this, {
+      select: buildSyncLookupSelect(this, {
```
Comment (tcodling) [inline packages/database/src/models/MedicationAdministrationRecord.ts:562]: None of the models in this pr are guaranteed to be encounter linked so we don't use this function

File: packages/central-server/__tests__/sync/CentralSyncManager.test.js [hunk 5]
```diff
@@ -3980,52 +3980,106 @@ describe('CentralSyncManager', () => {
         expect(updatedEncounterIds).not.toContain(encounter.id);
       });
 
-      it('will sync prescriptions linked through patient_ongoing_prescriptions if marked for sync', async () => {
```
Comment (tcodling) [inline packages/central-server/__tests__/sync/CentralSyncManager.test.js:3983]: This was actually not testing the edge case I intended to

File: packages/database/src/models/PatientOngoingPrescription.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (tcodling) [reply packages/database/src/models/PatientOngoingPrescription.ts:50]: seems like we would also need to tweak `MedicationAdministrationRecord` and `MedicationAdministrationRecordDose`

[the remark this answers is not in the document]
Comment (rohan-bes) [reply packages/database/src/models/PatientOngoingPrescription.ts:48]: Okay so having dug deeper there's never a situation where a prescription is linked to multiple encounter_prescriptions, so these left joins are safe :)

**Review by chris-bes — APPROVED** [review]
